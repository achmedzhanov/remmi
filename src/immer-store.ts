import produce from "immer"

type Handler<T = any> = (value: T) => void
type Selector<T = any, X = any> = (base: T) => X
type Disposer = () => void

interface Lens<T> {
    // static create
    select<X = any>(selector: (state: T) => X): Lens<X> // TODO: string based selector
    get(): T
    update(producer: ((draft: T) => void)): void // TODO: partial state
    subscribe(handler: Handler<T>): Disposer
    // selectAll(selector: string | ((state) => any)) // TODO type selector and such
    // merge(...lenses)
}

class Store<T = any> implements Lens<T> {
    state: T
    readonly subscriptions: Handler<T>[] = []

    constructor(initialValue: T) {
        this.state = initialValue
    }

    get() {
        return this.state
    }

    update(updater: ((draft: T) => T | void)) {
        const currentState = this.state
        // TODO support partial or new object?
        this.state = produce(this.state, updater)
        if (this.state !== currentState) notify(this.subscriptions, this.state)
    }

    subscribe(handler: Handler<T>) {
        return subscribe(this.subscriptions, handler)
    }

    select<X = any>(selector: Selector<T, X>): Select<X> {
        return new Select<X>(this, selector)
    }
}

class Select<T = any> implements Lens<T> {
    state?: T
    readonly subscriptions: Handler<T>[] = []
    hot = false
    parentSubscription?: Disposer

    constructor(private base: Lens<any>, private selector: Selector<any, T>) {}

    get() {
        if (!this.hot) return this.selector(this.base.get())
        return this.state!
    }

    update(updater: ((draft: T) => T | void)) {
        this.base.update(draft => {
            updater(this.selector(draft))
        })
    }

    subscribe(handler: Handler<T>) {
        if (!this.hot) this.resume()
        const disposer = subscribe(this.subscriptions, handler)
        return () => {
            disposer()
            if (this.subscriptions.length === 0) this.suspend()
        }
    }

    private resume() {
        this.hot = true
        this.parentSubscription = this.base.subscribe(nextBase => {
            const currentState = this.state
            this.state = this.selector(nextBase)
            if (this.state !== currentState)
                notify(this.subscriptions, this.state)
        })
    }

    private suspend() {
        if (this.hot) {
            this.parentSubscription!()
            this.parentSubscription = undefined
        }
        this.hot = false
    }

    select<X = any>(selector: Selector<T, X>): Select<X> {
        return new Select<X>(this, selector)
    }
}

function notify(subscriptions: Handler[], value: any) {
    subscriptions.forEach(f => f(value)) // optimize
}

function subscribe(subscriptions: Handler[], handler: Handler): Disposer {
    subscriptions.push(handler)
    return () => {
        const idx = subscriptions.indexOf(handler)
        if (idx !== -1) subscriptions.splice(idx, 1)
    }
}

export function createStore<T>(initialValue: T): Lens<T> {
    return new Store(initialValue)
}

// TODO refSelect

// function autoLens(initial) {
//     const store = new Store(initial)
//     return new Proxy(
//         {
//             $lens: store
//         },
//         traps
//     )
// }

// const traps = {
//     get(target, property) {
//         const value = target.get()[property]
//         if (isPrimitive(value)) return value
//         // todo cache lens in weakmap
//         return target.$lens.select(b => b[property])
//     },
//     set() {
//         throw new Error("Cannot write property, use 'update' instead")
//     }
//     // TODO: get prop descriptors from target.$lens.get..
// }

// function tracker(component) {
//     return class Tracker extends Component {
//         derivedPropsFromState(props, prevState) {
//             const oldLens = this.$lenses
//             const nextState = {}
//             const newLenses = Object.keys(props)
//                 .filter(prop => isLens(props[prop]))
//                 .map(prop => {
//                     const lens = props[prop]
//                     nextState[prop] = lens.get()
//                     return lens.subscribe(value => {
//                         setState(state => ({...state, [prop]: value}))
//                     })
//                 })

//             // TODO: could also reconcile lenses!
//             forEach(oldLens).unsubscribe()
//             return {...prevState, $lenses: newLenses}
//         }

//         render() {
//             const nextProps = {...props, ...state}
//             delete nextProps.lenses
//             return createElement(component, nextProps)
//         }
//     }
// }
