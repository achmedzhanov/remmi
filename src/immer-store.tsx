import produce from 'immer'
import * as React from 'react' // TODO: seperate bundle
import "./lenses/BaseLens";

import { Store } from './lenses/Store';
import { Lens, Selector } from './lenses/Lens';
import { Merge } from './lenses/Merge';

export function createStore<T>(initialValue: T): Lens<T> {
    return new Store(initialValue)
}

export function autoLens(baseLens: Lens) {
    return createProxyLens(baseLens, x => x)
}

// TODO proper typings
export function merge(...lenses: Lens[]) {
    return new Merge(lenses)
}

function createProxyLens(baseLens: Lens, property: string | number) {
    const lens = baseLens.select(property) // TODO: create delegator around lens to minimize api surface?
    return new Proxy(lens, traps)
}

const traps = {
    get(target: any, property: PropertyKey): any {
        if (property in target) {
            const value = target[property]
            if (typeof value === 'function')
                // TODO: blegh
                return value.bind(target) // avoid being called with 'this' as proxy
            return value
        }
        // optimize: cache
        return createProxyLens(target, property)
    },
    set() {
        throw new Error("Cannot write property, use 'update' instead")
    }
}

// TODO: read only lens

function isLens(thing: any): thing is Lens {
    return true // TODO
}

// TODO: type the thing
class Projector extends React.Component<
    {
        lenses: Lens[]
        children: any // can be type, if Projector would be exported
        // TODO: test if fresh closures cause invalidation or not
    },
    any
> {
    // TODO: type
    lastRenderedMerge
    subscription
    state = {}

    static getDerivedStateFromProps(props, state) {
        if (!state || !shallowEqual(state.lenses, props.lenses)) {
            const m = merge.apply(null, props.lenses)
            return {
                merge: m,
                lenses: props.lenses,
                props: m.value()
            }
        }
        return null
    }

    // shouldComponentUpdate(nextProps, nextState) {
    //     if (shallowEqual(nextProps.lenses, this.props.lenses) && shallowEqual(nextState.props, this.state.props)) {
    //         console.log('sCU false') // TODO: check if children changed?
    //         return false
    //     }
    //     return true
    // }

    render() {
        return this.props.children(this.state.props)
    }

    componentDidMount() {
        this.componentDidUpdate()
    }

    componentDidUpdate() {
        if (this.lastRenderedMerge !== this.state.merge) {
            this.lastRenderedMerge = this.state.merge
            const oldsub = this.subscription
            this.subscription = this.state.merge.subscribe(values => {
                this.setState(s => ({ ...s, props: values }))
            })
            oldsub && oldsub()
        }
    }

    componentWillUnmount() {
        this.subscription && this.subscription()
    }
}

class MapProjector extends React.Component<
    {
        lens: Lens
        children: any // can be type, if Projector would be exported
        // TODO: test if fresh closures cause invalidation or not
    },
    any
> {
    subscription
    lastRenderedLens

    state = {}

    render() {
        const { lens } = this.props
        const value = lens.value()
        console.dir(value)
        if (value === undefined || value === null) return null
        if (Array.isArray(value)) {
            return (
                <React.Fragment>
                    {value.map((v, idx) => {
                        console.dir(v)
                        return this.props.children(v, idx, lens.select(idx))
                    })}
                </React.Fragment>
            )
        }
        if (typeof value === 'object') {
            return (
                <React.Fragment>
                    {Object.keys(value).map(key =>
                        this.props.children(value[key], key, lens.select(key))
                    ) /* TODO: for key lenses should have unique id */}
                </React.Fragment>
            )
        }
        console.warn('projectMap expects lens to return an array, object or undefined')
        return null
    }

    componentDidMount() {
        this.componentDidUpdate()
    }

    componentDidUpdate() {
        if (this.lastRenderedLens !== this.props.lens) {
            console.log('subscribe')
            this.lastRenderedLens = this.props.lens
            const oldsub = this.subscription
            this.subscription = this.props.lens.subscribe(value => {
                console.log('update')
                // TODO: this.setState(s => ({ ...s, value }))
                this.forceUpdate()
            })
            oldsub && oldsub()
        }
    }

    componentWillUnmount() {
        this.subscription && this.subscription()
    }
}

// TODO: type
// TODO: also needs single arg overload
export function project(...args: any[]) {
    const children = args.pop()
    return React.createElement(
        Projector,
        {
            lenses: args
        },
        children
    )
}

export function mapProject(lens: Lens, callback: (value: any, idx: string | number, lens: Lens) => React.ReactNode) {
    return <MapProjector lens={lens}>{callback}</MapProjector>
}
