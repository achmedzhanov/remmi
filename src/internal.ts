import "./lenses/Lens" // without this rollup will drop the side-effect free Lens file
export * from "./lenses/Lens"
export * from "./lenses/BaseLens"
export * from "./lenses/Pipe"
export * from "./lenses/All"
export * from "./lenses/Merge"
export * from "./lenses/ReadOnly"
export * from "./lenses/Recorder"
export * from "./lenses/Select"
export * from "./lenses/SelectField"
export * from "./lenses/Store"
export * from "./lenses/ShallowEqual"
export * from "./react"
export * from "./utils"
export * from "./api"
export * from "./react"
export * from "./Tracker"