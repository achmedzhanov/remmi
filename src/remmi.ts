export { project, mapProject, createStore, merge, isLens, autorun, Tracker, Project }  from "./internal"
