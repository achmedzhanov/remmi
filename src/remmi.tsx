import produce from 'immer'
import React from 'react' // TODO: seperate bundle
import "./lenses/BaseLens";

import { Store } from './lenses/Store';
import { Lens, Selector } from './lenses/Lens';
import { Merge } from './lenses/Merge';

export function createStore<T>(initialValue: T): Lens<T> {
    return new Store(initialValue)
}

// TODO proper typings
export function merge(...lenses: Lens[]) {
    return new Merge(lenses)
}

// TODO: read only lens

function isLens(thing: any): thing is Lens {
    return true // TODO
}

export { project, mapProject }  from "./react"