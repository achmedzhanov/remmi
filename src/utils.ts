export const emptyArray = []
Object.freeze(emptyArray)

export function noop() {

}

export function fail(msg: string): never {
    throw new Error("[remmi] " + msg)
}

export function shallowEqual(ar1: any, ar2: any) {
    // TODO: also for objects
    if (ar1 === ar2) return true
    if (!ar1 || !ar2) return false
    if (ar1.length !== ar2.length) return false
    for (let i = 0; i < ar1.length; i++) if (ar1[i] !== ar2[i]) return false
    return true
}

export function once(fn: () => void): () => void {
    let executed = false
    return () => {
        if (!executed) {
            executed = true
            fn()
        }
    }
}