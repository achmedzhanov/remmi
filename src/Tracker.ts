import { Lens, Merge, shallowEqual, noop, Disposer } from "./internal";

const readListeners = new Set<(lens: Lens) => void>()

export function notifyRead(lens: Lens) {
    readListeners.forEach(h => h(lens)) // optimize
}

export class Tracker {
    merge: Merge = new Merge([])
    disposeMerge: Disposer = this.merge.subscribe(noop)

    constructor(private onInvalidate: () => void) {

    }

    public track<T>(fn: () => T): T {
        const dependencies = new Set<Lens>()
        const readListener = dependencies.add.bind(dependencies)
        readListeners.add(readListener)
        try {
            return fn()
        } finally {
            const newDeps = Array.from(dependencies)
            if (!shallowEqual(newDeps, this.merge.bases)) {
                const { disposeMerge } = this
                // optimization; don't create merge if only one dep
                this.merge = new Merge(newDeps as any) // TODO: fix typings
                this.disposeMerge = this.merge.subscribe(this.onInvalidate)
                disposeMerge()
            }
            readListeners.delete(readListener)
        }
    }

    public dispose() {
        this.disposeMerge()
    }
}

export function autorun(fn: () => void): Disposer {
    const t = new Tracker(() => {
        t.track(fn)
    })
    const res = t.dispose.bind(t)
    t.track(fn)
    return res
}