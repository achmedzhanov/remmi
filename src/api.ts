import {Lens, Store, BaseLens, AutoRender} from "./internal"
import * as React from "react";

let storeId = 0

export function createStore<T>(initialValue: T, name: string = "Store" + (++storeId)): Lens<T> {
    return new Store(initialValue, name)
}

// TODO: read only lens
export function isLens(thing: any): thing is Lens {
    return thing instanceof BaseLens
}

export function autoRender(fn: () => React.ReactNode) {
    return React.createElement(AutoRender, {}, fn)
}
