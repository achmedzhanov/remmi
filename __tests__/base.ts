"use strict"

import {createStore, autoLens, merge} from "../src/immer-store"

test("read & update through lens", () => {
    const data = {
        loc: {
            x: 3,
            y: 5
        }
    }

    const store = createStore(data)
    expect(store.value()).toBe(data)
    const lens = store.select(s => s.loc)

    expect(lens.value().x).toBe(3)
    const base = lens.value()
    lens.update(l => {
        l.x += 1
    })

    expect(lens.value()).not.toBe(base)
    expect(lens.value().x).toBe(4)

    expect(store.value()).not.toBe(data)
})

test("read & update through subscription", () => {
    const data = {
        loc: {
            x: 3,
            y: 5
        }
    }

    const values: any[] = []
    const store = createStore(data)
    const lens = store.select(s => s.loc)
    const d = lens.subscribe(next => {
        values.push(next)
    })

    lens.update(l => {
        l.x = 4
    })
    expect(lens.value()).toEqual({x: 4, y: 5})
    store.update(s => {
        s.loc.y = 6
    })
    store.update(s => {
        s.irrelevant = true
    })
    store.update(s => {
        s.loc = {a: 2}
    })
    store.update(s => {
        s.loc = null
    })
    store.update(s => {
        s.loc = {b: 3}
    })
    d()

    store.update(s => {
        s.loc = {c: 3}
    })
    expect(lens.value()).toEqual({c: 3})
    expect(values).toEqual([{x: 4, y: 5}, {x: 4, y: 6}, {a: 2}, null, {b: 3}])
})

test("read & update through proxy", () => {
    const data = {
        loc: {
            x: 3,
            y: 5
        }
    }

    const store = autoLens(createStore(data))
    expect(store.value()).toBe(data)
    const lens = store.loc

    expect(lens.x.value()).toBe(3)
    const base = lens.value()
    lens.update(l => {
        l.x += 1
    })

    expect(lens.value()).not.toBe(base)
    expect(lens.x.value()).toBe(4)

    expect(store.value()).not.toBe(data)
})

test("combine lenses", () => {
    const data = {
        users: {
            michel: {
                name: "michel",
                friend: "jan"
            },
            jan: {
                age: 10
            },
            piet: {
                age: 20
            }
        }
    }

    const ages = []
    const store = createStore(data)
    const michel = store.select(s => s.users.michel)
    const merger = merge(store, michel)
    // const friend = merger.select(([store, michel]) => store.users[michel.friend])
    const friend = merge(store.select(s => s.users), michel.select(m => m.friend)).select(([users, friend]) => users[friend])
    const age = friend.select(f => f.age)

    expect(friend.value().age).toBe(10)
    age.subscribe(age => ages.push(age))

    store.update(s => {
        s.users.jan.age = 12
    })
    store.update(s => {
        s.users.jan.age = 13
    })
    michel.update(m => {
        m.friend = "piet"
    })

    expect(friend.value()).toBe(store.value().users.piet)

    merger.update(([store, michel]) => {
        store.users.piet.age = 42
    })

    friend.update(f => {
        f.age = 43
    })

    expect(ages).toEqual([12, 13, 20, 42, 43])

    expect(store.value()).toEqual({
        users: {
            michel: {
                name: "michel",
                friend: "piet"
            },
            jan: {
                age: 13
            },
            piet: {
                age: 43
            }
        }
    })

    // TODO: test with proxy

    // TODO: test with  undefined

})

test("future ref", () => {
    const s = createStore({} as any)

    const values:any = []
    const michel = s.select(s => s.users).select(users => users && users.michel)
    michel.subscribe(v => values.push(v))

    s.update(s => {
        s.users = {}
    })
    s.update(s => {
        s.users["michel"] = "michel"
    })

    expect(values).toEqual(["michel"])
    expect(s.value()).toEqual({ users: { michel: "michel" }})
})