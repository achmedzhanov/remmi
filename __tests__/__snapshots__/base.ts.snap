// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combine lenses - fields 1`] = `
"Lens[store 
	.do(\\"users\\")
	.do(merge(store .do(\\"users\\").do(\\"michel\\").do(\\"friend\\")))
	.do(select(function (_a) { var users = _a[0],friend = _a[1]; return users[friend]; }))
]"
`

exports[`logging 1`] = `Array []`

exports[`logging 2`] = `
Array [
  Array [
    2,
    undefined,
    "Lens[test
	.do(\\"x\\")
	.do(\\"y\\")
	.do(log())
]",
  ],
  Array [
    2,
    undefined,
    "Lens[test
	.do(\\"x\\")
	.do(\\"y\\")
	.do(log())
]",
  ],
]
`

exports[`logging 3`] = `
Array [
  Array [
    2,
    undefined,
    "Lens[test
	.do(\\"x\\")
	.do(\\"y\\")
	.do(log())
]",
  ],
  Array [
    2,
    undefined,
    "Lens[test
	.do(\\"x\\")
	.do(\\"y\\")
	.do(log())
]",
  ],
  Array [
    3,
    undefined,
    "Lens[test
	.do(\\"x\\")
	.do(\\"y\\")
	.do(log())
]",
  ],
]
`
